### Q1 Pandas version

In [29]:
import pandas as pd

pd.__version__

'2.3.2'

**Answer**: Pandas version in use is 2.3.2.

### Q2 Records count

In [30]:
df = pd.read_csv("../data/car_fuel_efficiency.csv")
# First let's have a peek what we have in the file.
df.head()

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                  170            3.0       159.0     3413.433759   
1                  130            5.0        97.0     3149.664934   
2                  170            NaN        78.0     3079.038997   
3                  220            4.0         NaN     2542.392402   
4                  210            1.0       140.0     3460.870990   

   acceleration  model_year  origin fuel_type         drivetrain  num_doors  \
0          17.7        2003  Europe  Gasoline    All-wheel drive        0.0   
1          17.8        2007     USA  Gasoline  Front-wheel drive        0.0   
2          15.1        2018  Europe  Gasoline  Front-wheel drive        0.0   
3          20.2        2009     USA    Diesel    All-wheel drive        2.0   
4          14.4        2009  Europe  Gasoline    All-wheel drive        2.0   

   fuel_efficiency_mpg  
0            13.231729  
1            13.688217  
2            14.246341  
3            16.912736  
4            12.488369

In [31]:
records = len(df)  # Prints number of records.
print(records)

9704


**Answer**: There is 9704 records in the csv file.

### Q3 Fuel types

In [32]:
fuel = df[
    "fuel_type"
].nunique()  # This print number of unique values in fuel_type column.
print(fuel)

2


**Answer**: There is only two distinct values in the fuel types.

### Q4 Missing values

In [33]:
missing = df.isnull().sum().sum()
print(missing)

2622


**Answer**: Total amount of missing values is 2622.

### Q5 Max fuel efficiency

In [34]:
# Fuel efficiency, cars from asia
cars_asia = df.loc[df["origin"] == "Asia", "fuel_efficiency_mpg"]

# Find the maximum value from the filtered data
max_asia = cars_asia.max()
print(max_asia)

23.759122836520497


**Answer**: Max fuel efficiency in a car made in Asia was approx 23,8.

### Q6 Median value of horsepower

In [35]:
# 1 median of horsepower
horsepow = df["horsepower"].median()
print(horsepow)

149.0


In [36]:
# 2 Find most frequent value in horsepower
frequent = df["horsepower"].mode()
print(frequent)

0    152.0
Name: horsepower, dtype: float64


In [37]:
# 3 Fill missing values with most frequent one
df["horsepower"] = df["horsepower"].fillna(frequent)

In [38]:
# 4 median of horsepower after filling
horsepow_filled = df["horsepower"].median()
print(horsepow_filled)

149.0


**Answer**: Has the value changed? No, median didn't change in this case.

### Q7 Sum of weights